### DISCUSSION TITLE
Django: A Serverless Function has exceeded the unzipped maximum size of 250 MB.

### DISCUSSION THREAD

**Discussion by Jasiri-w** [vercel/community#3381] [Help] [answered]
### Summary

I am migrating from Railway and deploying my Django application straight from GitHub on Vercel and I'm experiencing this error claiming that "A Serverless Function has exceeded the unzipped maximum size of 250 MB." What can I do to debug, find the specific function that is causing this issue and reduce its size?

Why not use solutions from other threads? Unfortunately because the response had no reference to a specific function I do not know where to start. I would also like to know how I could investigate that.

![image](https://github.com/vercel/community/assets/6060064/18767277-45f7-4428-a5de-43aa4ff91bda)

requirements.txt:
anyio==3.6.1
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
asgiref==3.6.0
asttokens==2.0.8
attrs==22.1.0
Babel==2.10.3
backcall==0.2.0
beautifulsoup4==4.11.1
bleach==5.0.1
certifi==2022.6.15.1
cffi==1.15.1
charset-normalizer==2.1.1
colorama==0.4.5
cryptography==40.0.1
debugpy==1.6.3
decorator==5.1.1
defusedxml==0.7.1
Django==4.2
django-allauth==0.54.0
djangorestframework==3.14.0
entrypoints==0.4
executing==1.0.0
fastjsonschema==2.16.1
gunicorn==20.1.0
idna==3.3
ipykernel==6.15.2
ipython==8.5.0
ipython-genutils==0.2.0
jedi==0.18.1
Jinja2==3.1.2
json5==0.9.10
jsonschema==4.16.0
lxml==4.9.1
MarkupSafe==2.1.1
matplotlib-inline==0.1.6
mistune==0.8.4
nbclassic==0.4.3
nbclient==0.5.13
nbconvert==6.5.3
nbformat==5.5.0
nest-asyncio==1.5.5
notebook==6.4.12
notebook-shim==0.1.0
oauthlib==3.2.2
packaging==21.3
pandocfilters==1.5.0
parso==0.8.3
pickleshare==0.7.5
prometheus-client==0.14.1
prompt-toolkit==3.0.31
psutil==5.9.2
pure-eval==0.2.2
pycparser==2.21
Pygments==2.13.0
PyJWT==2.6.0
pyparsing==3.0.9
pyrsistent==0.18.1
python-dateutil==2.8.2
python3-openid==3.2.0
pytz==2022.2.1
PyYAML==6.0
pyzmq==23.2.1
requests==2.28.1
requests-oauthlib==1.3.1
Send2Trash==1.8.0
six==1.16.0
sniffio==1.3.0
soupsieve==2.3.2.post1
sqlparse==0.4.3
stack-data==0.5.0
terminado==0.15.0
tinycss2==1.1.1
tomli==2.0.1
tornado==6.2
traitlets==5.4.0
tzdata==2023.3
urllib3==1.26.12
voila==0.3.6
wcwidth==0.2.5
webencodings==0.5.1
websocket-client==1.4.1
websockets==10.3
whitenoise==6.2.0

### Example

https://github.com/Jasiri-w/Sportshall-Registration-Web-App

### Steps to Reproduce

I have no idea, though you may look through the repo linked above to see where perhaps I could make changes.

**Comment by abdelkareemkobo** [comment]
did you solve it !

**Reply by Jasiri-w** [reply]
Nope ! Still desperate for some assistance

**Reply by cannonmaster** [reply]
same happen to me, do you guys found a solution on it?

**Reply by zhyd1997** [reply]
please check deps folder size and see which dep size is too large.

```shell
# MacOS Terminal
du -h -d 1 /path/to/folder | sort -h
```

**Reply by Jasiri-w** [reply]
Ok I have ran this command adding * as a wildcard at the end of the path to look at every single subdirectory, yet there are no folders exceeding 5MB which is a far cry from the 250MB limit. How can I single out the specific folder causing my Serverless functions to be too large?

**Reply by zhyd1997** [reply]
Please check every 1st layer subfolders in `.venv/lib` folder, which includes your all project deps, I promise you will find the folders exceeding **5MB**.

**Comment by Jasiri-w** [comment, marked as answer]
# Solved

My issue was with dependencies. I used a brute force ground-up approach to figure this out by deleting my requirements.txt file and reinstalling all the necessary packages one by one in an isolated virtual environment.

## Why did this happen:
I disregarded the proper programming convention and did not use a virtual environment when installing packages through pip. I also used several different machines to work on my project, one of which was a VS Code programming environment straight on the GitHub repository. So when the time came to "pip freeze" all my dependencies into my  requirements file, many irrelevant packages that were intrinsic to those machines including things like Jupyter-lab contaminated my python files inflating my builds.

**Reply by alvindera97** [reply]
> please check deps folder size and see which dep size is too large.
> 
> ```shell
> # MacOS Terminal
> du -h -d 1 /path/to/folder | sort -h
> ```



Here's what I found out: 

```
du -h -d 1 ../lib/python3.10/site-packages/ | sort -h
12K	../lib/python3.10/site-packages/google
20K	../lib/python3.10/site-packages/dj_database_url
24K	../lib/python3.10/site-packages/et_xmlfile
28K	../lib/python3.10/site-packages/certifi-2023.7.22.dist-info
28K	../lib/python3.10/site-packages/_distutils_hack
28K	../lib/python3.10/site-packages/permissionedforms
28K	../lib/python3.10/site-packages/six-1.16.0.dist-info
28K	../lib/python3.10/site-packages/sqlparse-0.4.4.dist-info
28K	../lib/python3.10/site-packages/telepath-0.3.1.dist-info
28K	../lib/python3.10/site-packages/whitenoise-6.6.0.dist-info
28K	../lib/python3.10/site-packages/willow-1.6.2.dist-info
32K	../lib/python3.10/site-packages/beautifulsoup4-4.11.2.dist-info
32K	../lib/python3.10/site-packages/django_filter-23.3.dist-info
32K	../lib/python3.10/site-packages/django_modelcluster-6.1.dist-info
32K	../lib/python3.10/site-packages/django_permissionedforms-0.1.dist-info
32K	../lib/python3.10/site-packages/et_xmlfile-1.1.0.dist-info
32K	../lib/python3.10/site-packages/idna-3.4.dist-info
32K	../lib/python3.10/site-packages/jmespath-1.0.1.dist-info
32K	../lib/python3.10/site-packages/l18n-2021.3.dist-info
32K	../lib/python3.10/site-packages/psycopg2_binary-2.9.9.dist-info
32K	../lib/python3.10/site-packages/soupsieve-2.5.dist-info
32K	../lib/python3.10/site-packages/webencodings-0.5.1.dist-info
32K	../lib/python3.10/site-packages/wheel-0.41.2.dist-info
36K	../lib/python3.10/site-packages/asgiref-3.7.2.dist-info
36K	../lib/python3.10/site-packages/django_ipware-5.0.2.dist-info
36K	../lib/python3.10/site-packages/django_taggit-4.0.0.dist-info
36K	../lib/python3.10/site-packages/dj_database_url-2.1.0.dist-info
36K	../lib/python3.10/site-packages/typing_extensions-4.8.0.dist-info
36K	../lib/python3.10/site-packages/urllib3-2.0.7.dist-info
40K	../lib/python3.10/site-packages/django_treebeard-4.7.dist-info
40K	../lib/python3.10/site-packages/filetype-1.2.0.dist-info
40K	../lib/python3.10/site-packages/pillow_heif-0.13.1.dist-info
40K	../lib/python3.10/site-packages/python_dateutil-2.8.2.dist-info
40K	../lib/python3.10/site-packages/python_decouple-3.8.dist-info
40K	../lib/python3.10/site-packages/requests-2.31.0.dist-info
40K	../lib/python3.10/site-packages/s3transfer-0.7.0.dist-info
44K	../lib/python3.10/site-packages/draftjs_exporter-2.1.7.dist-info
48K	../lib/python3.10/site-packages/boto3-1.28.73.dist-info
48K	../lib/python3.10/site-packages/html5lib-1.1.dist-info
52K	../lib/python3.10/site-packages/openpyxl-3.1.2.dist-info
56K	../lib/python3.10/site-packages/defusedxml-0.7.1.dist-info
56K	../lib/python3.10/site-packages/ipware
56K	../lib/python3.10/site-packages/setuptools-59.6.0.dist-info
64K	../lib/python3.10/site-packages/charset_normalizer-3.3.1.dist-info
64K	../lib/python3.10/site-packages/djangorestframework-3.14.0.dist-info
68K	../lib/python3.10/site-packages/anyascii-0.3.2.dist-info
80K	../lib/python3.10/site-packages/Pillow-10.1.0.dist-info
88K	../lib/python3.10/site-packages/django_storages-1.14.2.dist-info
88K	../lib/python3.10/site-packages/webencodings
92K	../lib/python3.10/site-packages/telepath
100K	../lib/python3.10/site-packages/pytz-2023.3.post1.dist-info
120K	../lib/python3.10/site-packages/defusedxml
120K	../lib/python3.10/site-packages/pip-23.3.1.dist-info
128K	../lib/python3.10/site-packages/__pycache__
152K	../lib/python3.10/site-packages/asgiref
156K	../lib/python3.10/site-packages/jmespath
176K	../lib/python3.10/site-packages/whitenoise
184K	../lib/python3.10/site-packages/pillow_heif
192K	../lib/python3.10/site-packages/draftjs_exporter
204K	../lib/python3.10/site-packages/modelcluster
216K	../lib/python3.10/site-packages/filetype
228K	../lib/python3.10/site-packages/botocore-1.31.73.dist-info
244K	../lib/python3.10/site-packages/storages
256K	../lib/python3.10/site-packages/l18n
260K	../lib/python3.10/site-packages/soupsieve
300K	../lib/python3.10/site-packages/tests
312K	../lib/python3.10/site-packages/certifi
356K	../lib/python3.10/site-packages/sqlparse
400K	../lib/python3.10/site-packages/requests
512K	../lib/python3.10/site-packages/wheel
516K	../lib/python3.10/site-packages/idna
524K	../lib/python3.10/site-packages/Django-4.2.6.dist-info
536K	../lib/python3.10/site-packages/taggit
544K	../lib/python3.10/site-packages/wagtail-5.1.3.dist-info
564K	../lib/python3.10/site-packages/charset_normalizer
572K	../lib/python3.10/site-packages/s3transfer
616K	../lib/python3.10/site-packages/psycopg2
624K	../lib/python3.10/site-packages/django_filters
748K	../lib/python3.10/site-packages/dateutil
832K	../lib/python3.10/site-packages/treebeard
848K	../lib/python3.10/site-packages/urllib3
960K	../lib/python3.10/site-packages/bs4
972K	../lib/python3.10/site-packages/pkg_resources
1.1M	../lib/python3.10/site-packages/html5lib
1.1M	../lib/python3.10/site-packages/willow
1.5M	../lib/python3.10/site-packages/boto3
2.4M	../lib/python3.10/site-packages/anyascii
2.6M	../lib/python3.10/site-packages/openpyxl
2.8M	../lib/python3.10/site-packages/pytz
3.0M	../lib/python3.10/site-packages/PIL
4.0M	../lib/python3.10/site-packages/setuptools
4.8M	../lib/python3.10/site-packages/rest_framework
6.8M	../lib/python3.10/site-packages/psycopg2_binary.libs
9.1M	../lib/python3.10/site-packages/Pillow.libs
15M	../lib/python3.10/site-packages/pip
31M	../lib/python3.10/site-packages/pillow_heif.libs
46M	../lib/python3.10/site-packages/django
53M	../lib/python3.10/site-packages/wagtail
86M	../lib/python3.10/site-packages/botocore
282M	../lib/python3.10/site-packages/
```

**So Vercel's message of exceeding 250MB checks out because I'm seeing a grand whole total of 282MB**.....  with `botocore` having a whooping **86MB**! Right now I have to look for alternatives to botocore with a lot smaller size. Or if there was a way to just clean out modules and methods you simply don't need from a package for your application to execute 😪.

I'm still asking myself if https://avilpage.com/2020/02/reduce-python-package-footprint.html will be a workaround worth considering. Maybe I could manually take out a lot of what's not being called by my source code.

Maybe [TreeShaking](https://en.wikipedia.org/wiki/Tree_shaking) for Python might be the way.

**Reply by alvindera97** [reply]
**UPDATE**

This sort of solved my problems: https://blog.nathanv.me/posts/stripping-botocore-python-serverless/

**Comment by ashraymallesh** [comment]
If anyone sees this in the future, this worked for me:

https://github.com/orgs/vercel/discussions/103#discussioncomment-6952635

try `vercel build` locally and look at the size of the output folder inside `.vercel`. It's definitely over 250mb and most of it is not your application. It's the .next folder. This vercel.json excludes those folders and you should now have an output under 250mb.
